pywr/model.py

```diff
@@ -1 +1,154 @@
+            data = data.read()
+            return cls.loads(data, model, path, solver)
+
+        return cls._load_from_dict(data, model=model, path=path, solver=None, **kwargs)
+
+    @classmethod
+    def _load_from_dict(cls, data, model=None, path=None, solver=None, **kwargs):
+        """Load data from a dictionary."""
+        # data is a dictionary, make a copy to avoid modify the input
+        data = copy.deepcopy(data)
+
+        if model is None:
+            model = cls()
+
+        timestepper_data = data["timestepper"]
+
+        # Load the sub-models
+        sub_model_paths = data["models"]
+        for sub_model_definition in sub_model_paths:
+            sub_model_name = sub_model_definition["name"]
+            sub_model_path = sub_model_definition["path"]
+            sub_model_solver = sub_model_definition.get("solver", solver)
+            if path is not None:
+                sub_model_path = os.path.join(os.path.dirname(path), sub_model_path)
+
+            sub_model = Model.load(sub_model_path, path=path, solver=sub_model_solver)
+            sub_model.timestepper.start = timestepper_data["start"]
+            sub_model.timestepper.end = timestepper_data["end"]
+            sub_model.timestepper.timestep = timestepper_data["timestep"]
+            model.add_model(sub_model_name, sub_model)
+
+        return model
+
+    @property
+    def dirty(self) -> bool:
+        """Returns true if any of the models are 'dirty'."""
+        return any(
+            model.dirty or model.timestepper.dirty for model in self.models.values()
+        )
+
+    def setup(self):
+        for model in self.models.values():
+            model.setup()
+
+    def reset(self):
+        for model in self.models.values():
+            model.reset()
+
+    # def _step(self):
+    #     for model in self.models.values():
+    #         model.before()
+    #     # solve the current timestep for all models
+    #     rets = []
+    #     for model in self.models.values():
+    #         ret = model.solve()
+    #         rets.append(ret)
+    #
+    #     for model in self.models.values():
+    #         model.after()
+    #     return rets
+
+    def _step(self):
+        rets = []
+        for model in self.models.values():
+            model.before()
+            ret = model.solve()
+            rets.append(ret)
+            model.after()
+        return rets
+
+    def run(self):
+        """Run the multi model simulation"""
+        logger.info("Start multi model run ...")
+        t0 = time.time()
+        timestep = None
+
+        # Get each model's timestepper
+        timesteppers = [m.timestepper for m in self.models.values()]
+
+        try:
+            if self.dirty:
+                self.setup()
+            else:
+                self.reset()
+            t1 = time.time()
+
+            # Begin time-stepping
+            for timesteps in zip(*timesteppers):
+                # TODO assert these are all the same date
```

Comment:
If the timesteppers are being set equal in the load method is an assert needed here?

Reply:
They can be modified between load and run, or also created without the load method. So, this check is a good one. It's now implemented.